boards.espressif.memory_layout.mcuboot fails in many boards

### Describe the bug

The newly introduced test boards.espressif.memory_layout.mcuboot fails in many targets:
 
* m5stack_core2/esp32/appcpu
* m5stack_fire/esp32/appcpu
* m5stack_atom_lite/esp32/appcpu
* dnesp32s3b/esp32s3/procpu
* m5stack_atoms3/esp32s3/appcpu
* doit_esp32_devkit_v1/esp32/appcpu
* dnesp32s3b/esp32s3/appcpu
* m5stack_atoms3_lite/esp32s3/appcpu
* m5stack_cores3/esp32s3/procpu
* m5stickc_plus/esp32/appcpu
* m5stack_cores3/esp32s3/appcpu
* m5stack_cores3/esp32s3/procpu/se
* m5stickc_plus/esp32/procpu
* rak3112/esp32s3/appcpu
* heltec_wireless_tracker/esp32s3/appcpu
* heltec_wireless_stick_lite_v3/esp32s3/appcpu
* heltec_wifi_lora32_v3/esp32s3/appcpu
* heltec_wifi_lora32_v2/esp32/appcpu
* heltec_wireless_stick_lite_v3/esp32s3/procpu
* heltec_wifi_lora32_v3/esp32s3/procpu
* esp32s3_eye/esp32s3/appcpu
* esp32_ethernet_kit/esp32/appcpu
* esp32s3_devkitc/esp32s3/appcpu
* esp_wrover_kit/esp32/appcpu
* esp32c3_rust/esp32c3
* esp32_devkitc/esp32/appcpu
* adafruit_feather_esp32@2/esp32/appcpu
* esp32_cam/esp32/appcpu
* adafruit_feather_esp32s3_tft/esp32s3/appcpu
* adafruit_qt_py_esp32s3/esp32s3/appcpu
* adafruit_qt_py_esp32s3@psram/esp32s3/appcpu
* adafruit_feather_esp32s3_tft_reverse/esp32s3/appcpu
* adafruit_feather_esp32s3/esp32s3/appcpu
* walter/esp32s3/appcpu
* esp32s3_luatos_core/esp32s3/appcpu
* esp32s3_luatos_core/esp32s3/appcpu/usb
* kincony_kc868_a32/esp32/appcpu
* kincony_kc868_a32/esp32/procpu
* xiao_esp32s3/esp32s3/appcpu
* lolin32_lite/esp32/appcpu
* ttgo_t7v1_5/esp32/appcpu
* tdongle_s3/esp32s3/appcpu
* ttgo_tbeam/esp32/appcpu
* ttgo_lora32/esp32/appcpu
* twatch_s3/esp32s3/appcpu
* ttgo_t8s3/esp32s3/appcpu
* yd_esp32/esp32/appcpu
* olimex_esp32_evb/esp32/appcpu
* esp32s3_touch_lcd_1_28/esp32s3/appcpu
* esp32s3_rlcd_4_2/esp32s3/appcpu
* weact_esp32s3_mini/esp32s3/appcpu
* weact_esp32s3_b/esp32s3/appcpu
* odroid_go/esp32/appcpu
* esp32s3_matrix/esp32s3/appcpu
* esp32s3_geek/esp32s3/appcpu

Secondary issue: The tests is filtering only kconfig which means it the cmake configuration step will be run for *all* targets to be able to filter. It would be much nicer if the filter could include an architecture filter so the majority of targets would be excluded from the beginning saving ~1h of CPU compute time.

### Regression

- [ ] This is a regression.

### Steps to reproduce

To get all failures: `twister -T tests/boards/espressif/memory_layout/ --all`
Just one failure:
```
mkdir build && cd build
cmake -GNinja -DBOARD=adafruit_feather_esp32s3_tft/esp32s3/appcpu -DAPP_DIR=../tests/boards/espressif/memory_layout/ ../share/sysbuild/
ninja
```

### Relevant log output

```shell
/opt/toolchains/zephyr-sdk-0.17.4/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: warning: cannot find entry symbol __start; not setting start address
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

- OS Ubuntu 24.04
- Zephyr SDK 0.17.4
- SHA: main since 87e1920f0970a4346093c298e5d2f58cda4fc19e

### Additional Context

Failed since introduced in
* https://github.com/zephyrproject-rtos/zephyr/pull/104216/

--- a/tests/boards/espressif/memory_layout/testcase.yaml
+++ b/tests/boards/espressif/memory_layout/testcase.yaml
@@ -1,6 +1,13 @@
 common:
   tags: memory
-  filter: CONFIG_SOC_FAMILY_ESPRESSIF_ESP32
+  platform_allow:
+    - esp32_devkitc/esp32/procpu
+    - esp32s2_devkitc
+    - esp32s3_devkitc/esp32s3/procpu
+    - esp8684_devkitm
+    - esp32c3_devkitm
+    - esp32c6_devkitc/esp32c6/hpcore
+    - esp32h2_devkitm
   harness: console
   harness_config:
     type: multi_line

